pkg/stanza/fileconsumer/internal/tracker/tracker.go

```diff
@@ -0,0 +1,200 @@
+func (t *Tracker) PreviousPollFiles() *fileset.Fileset[*reader.Reader] {
+	return t.previousPollFiles
+}
+
+func (t *Tracker) Persist(persister operator.Persister) {
+	if persister != nil {
+		allCheckpoints := make([]*reader.Metadata, 0, t.TotalReaders())
+		for _, knownFiles := range t.knownFiles {
+			allCheckpoints = append(allCheckpoints, knownFiles.Get()...)
+		}
+
+		for _, r := range t.previousPollFiles.Get() {
+			allCheckpoints = append(allCheckpoints, r.Metadata)
+		}
+		if err := checkpoint.Save(context.Background(), persister, allCheckpoints); err != nil {
+			t.Errorw("save offsets", zap.Error(err))
+		}
+	}
+}
+
+func (t *Tracker) Load(ctx context.Context, persister operator.Persister) error {
+	offsets, err := checkpoint.Load(ctx, persister)
+	if err != nil {
+		return fmt.Errorf("read known files from database: %w", err)
+	}
+	if len(offsets) > 0 {
+		t.Infow("Resuming from previously known offset(s). 'start_at' setting is not applicable.")
+		t.ReaderFactory.FromBeginning = true
+		t.knownFiles[0].Add(offsets...)
+	}
+	return nil
+}
+
+func (t *Tracker) ClosePreviousFiles() {
+	// t.previousPollFiles -> t.knownFiles[0]
+
+	for r, _ := t.previousPollFiles.Pop(); r != nil; r, _ = t.previousPollFiles.Pop() {
+		t.knownFiles[0].Add(r.Close())
+	}
+}
+
+func (t *Tracker) rotateFilesets() {
+	// shift the filesets at end of every poll() call
+	// t.knownFiles[0] -> t.knownFiles[1] -> t.knownFiles[2]
+	copy(t.knownFiles[1:], t.knownFiles)
+	t.knownFiles[0] = fileset.New[*reader.Metadata](t.maxBatchFiles / 2)
+}
+
+func (t *Tracker) makeFingerprint(path string) (*fingerprint.Fingerprint, *os.File) {
+	file, err := os.Open(path) // #nosec - operator must read in files defined by user
+	if err != nil {
+		t.Errorw("Failed to open file", zap.Error(err))
+		return nil, nil
+	}
+
+	fp, err := t.ReaderFactory.NewFingerprint(file)
+	if err != nil {
+		if err = file.Close(); err != nil {
+			t.Debugw("problem closing file", zap.Error(err))
+		}
+		return nil, nil
+	}
+
+	if len(fp.FirstBytes) == 0 {
+		// Empty file, don't read it until we can compare its fingerprint
+		if err = file.Close(); err != nil {
+			t.Debugw("problem closing file", zap.Error(err))
+		}
+		return nil, nil
+	}
+	return fp, file
+}
+
+func (t *Tracker) newReader(file *os.File, fp *fingerprint.Fingerprint) (*reader.Reader, error) {
+	// Check previous poll cycle for match
+	if oldReader := t.previousPollFiles.Match(fp, fileset.StartsWith); oldReader != nil {
+		return t.ReaderFactory.NewReaderFromMetadata(file, oldReader.Close())
+	}
+
+	// Iterate backwards to match newest first
+	for i := 0; i < len(t.knownFiles); i++ {
+		if oldMetadata := t.knownFiles[i].Match(fp, fileset.StartsWith); oldMetadata != nil {
+			return t.ReaderFactory.NewReaderFromMetadata(file, oldMetadata)
+		}
+	}
+
+	// If we don't match any previously known files, create a new reader from scratch
+	t.Infow("Started watching file", "path", file.Name())
+	return t.ReaderFactory.NewReader(file, fp)
+}
```

Comment:
I think this is where it gets tricky to cleanly separate the tracker while still feeling like the tracker is simplifying the code. However, I think it may be helpful to consider that eventually this tracker implementation should be one of several alternatives. So basically we should be thinking about it in terms of an interface which would allow other implementations (such as async).

Therefore, I think we should mostly move this back to manager and make a few things more abstract, so that the manager is doing something less specific to the implementation:

```suggestion
func (m *Manager) newReader(file *os.File, fp *fingerprint.Fingerprint) (*reader.Reader, error) {
	if oldReader := m.tracker.GetOpenFile(fp, fileset.StartsWith); oldReader != nil {
		return t.ReaderFactory.NewReaderFromMetadata(file, oldReader.Close())
	}
	if oldMetadata := m.tracker.GetClosedFile(fp, fileset.StartsWith); oldMetadata != nil {
		return t.ReaderFactory.NewReaderFromMetadata(file, oldMetadata)
	}
	m.Infow("Started watching file", "path", file.Name())
	return m.readerFactory.NewReader(file, fp)
}
```